File: blog.py
from flask import Blueprint, render_template, redirect, url_for, flash, request
from flask_login import login_required, current_user
from app import db
from models import BlogPost
import bleach

blog = Blueprint('blog', __name__)

ALLOWED_TAGS = [
    'p', 'br', 'strong', 'em', 'u', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6',
    'ul', 'ol', 'li', 'a', 'blockquote', 'code', 'pre'
]

ALLOWED_ATTRIBUTES = {
    'a': ['href', 'title'],
    '*': ['class']
}

@blog.route('/blog')
def index():
    posts = BlogPost.query.filter_by(published=True).order_by(BlogPost.created_at.desc()).all()
    return render_template('blog/index.html', posts=posts)

@blog.route('/blog/<slug>')
def post(slug):
    post = BlogPost.query.filter_by(slug=slug).first_or_404()
    if not post.published and not (current_user.is_authenticated and current_user.is_admin):
        flash('This blog post is not published yet.', 'warning')
        return redirect(url_for('blog.index'))
    return render_template('blog/post.html', post=post)

@blog.route('/admin/blog-posts')
@login_required
def blog_posts():
    if not current_user.is_admin:
        flash('Access denied. Admin rights required.', 'danger')
        return redirect(url_for('index'))
    
    posts = BlogPost.query.order_by(BlogPost.created_at.desc()).all()
    return render_template('admin/blog_posts.html', blog_posts=posts)

@blog.route('/blog/create', methods=['GET', 'POST'])
@login_required
def create_post():
    if not current_user.is_admin:
        flash('Access denied. Admin rights required.', 'danger')
        return redirect(url_for('index'))
    
    if request.method == 'POST':
        title = request.form.get('title')
        content = bleach.clean(request.form.get('content') or '', tags=ALLOWED_TAGS, attributes=ALLOWED_ATTRIBUTES)
        excerpt = request.form.get('excerpt')
        featured_image = request.form.get('featured_image')
        published = bool(request.form.get('published'))
        
        if not title or not content:
            flash('Title and content are required.', 'danger')
            return redirect(url_for('blog.create_post'))
        
        post = BlogPost()
        post.title = title
        post.content = content
        post.excerpt = excerpt
        post.featured_image = featured_image
        post.published = published
        post.user_id = current_user.id
        post.slug = post.generate_slug()
        
        try:
            db.session.add(post)
            db.session.commit()
            flash('Blog post created successfully!', 'success')
            return redirect(url_for('blog.blog_posts'))
        except Exception as e:
            db.session.rollback()
            flash(f'Error creating blog post: {str(e)}', 'danger')
            return redirect(url_for('blog.create_post'))
    
    return render_template('admin/blog_post_form.html', post=None)

@blog.route('/blog/edit/<int:post_id>', methods=['GET', 'POST'])
@login_required
def edit_post(post_id):
    if not current_user.is_admin:
        flash('Access denied. Admin rights required.', 'danger')
        return redirect(url_for('index'))
    
    post = BlogPost.query.get_or_404(post_id)
    
    if request.method == 'POST':
        title = request.form.get('title')
        content = bleach.clean(request.form.get('content') or '', tags=ALLOWED_TAGS, attributes=ALLOWED_ATTRIBUTES)
        excerpt = request.form.get('excerpt')
        featured_image = request.form.get('featured_image')
        published = bool(request.form.get('published'))
        
        if not title or not content:
            flash('Title and content are required.', 'danger')
            return redirect(url_for('blog.edit_post', post_id=post_id))
        
        try:
            post.title = title
            if title != post.title:
                post.slug = post.generate_slug()
            post.content = content
            post.excerpt = excerpt
            post.featured_image = featured_image
            post.published = published
            
            db.session.commit()
            flash('Blog post updated successfully!', 'success')
            return redirect(url_for('blog.blog_posts'))
        except Exception as e:
            db.session.rollback()
            flash(f'Error updating blog post: {str(e)}', 'danger')
            return redirect(url_for('blog.edit_post', post_id=post_id))
    
    return render_template('admin/blog_post_form.html', post=post)

@blog.route('/blog/delete/<int:post_id>')
@login_required
def delete_post(post_id):
    if not current_user.is_admin:
        flash('Access denied. Admin rights required.', 'danger')
        return redirect(url_for('index'))
    
    post = BlogPost.query.get_or_404(post_id)
    
    try:
        db.session.delete(post)
        db.session.commit()
        flash('Blog post deleted successfully!', 'success')
    except Exception as e:
        db.session.rollback()
        flash(f'Error deleting blog post: {str(e)}', 'danger')
    
    return redirect(url_for('blog.blog_posts'))

@blog.route('/blog/toggle/<int:post_id>')
@login_required
def toggle_post(post_id):
    if not current_user.is_admin:
        flash('Access denied. Admin rights required.', 'danger')
        return redirect(url_for('index'))
    
    post = BlogPost.query.get_or_404(post_id)
    
    try:
        post.published = not post.published
        db.session.commit()
        status = 'published' if post.published else 'unpublished'
        flash(f'Blog post {status} successfully!', 'success')
    except Exception as e:
        db.session.rollback()
        flash(f'Error toggling blog post status: {str(e)}', 'danger')
    
    return redirect(url_for('blog.blog_posts'))

File: routes/blog.js
const express = require('express');
const router = express.Router();
const { BlogPost, User } = require('../models');

// Blog index
router.get('/', async (req, res) => {
  try {
    const posts = await BlogPost.findAll({
      include: [
        {
          model: User,
          as: 'author',
          attributes: ['username']
        }
      ],
      order: [['createdAt', 'DESC']]
    });
    
    res.render('blog/index', { posts });
  } catch (error) {
    console.error('Blog index error:', error);
    res.render('error', { message: 'Failed to load blog posts' });
  }
});

// Blog post detail
router.get('/post/:id', async (req, res) => {
  try {
    const post = await BlogPost.findByPk(req.params.id, {
      include: [
        {
          model: User,
          as: 'author',
          attributes: ['username']
        }
      ]
    });
    
    if (!post) {
      return res.status(404).render('error', { message: 'Blog post not found' });
    }
    
    res.render('blog/post', { post });
  } catch (error) {
    console.error('Blog post error:', error);
    res.render('error', { message: 'Failed to load blog post' });
  }
});

// Admin blog posts list
router.get('/admin', requireAdmin, async (req, res) => {
  try {
    const posts = await BlogPost.findAll({
      include: [
        {
          model: User,
          as: 'author',
          attributes: ['username']
        }
      ],
      order: [['createdAt', 'DESC']]
    });
    
    res.render('admin/blog/blog_posts', { posts });
  } catch (error) {
    console.error('Admin blog posts error:', error);
    res.render('error', { message: 'Failed to load blog posts' });
  }
});

// Create blog post form
router.get('/admin/create', requireAdmin, (req, res) => {
  res.render('admin/blog_post_form');
});

// Create blog post
router.post('/admin/create', requireAdmin, async (req, res) => {
  try {
    const { title, content, excerpt } = req.body;
    
    if (!title || !content) {
      req.session.flash = { type: 'danger', message: 'Title and content are required' };
      return res.redirect('/blog/admin/create');
    }
    
    await BlogPost.create({
      title,
      content,
      excerpt: excerpt || content.substring(0, 200) + '...',
      user_id: req.session.user.id
    });
    
    req.session.flash = { type: 'success', message: 'Blog post created successfully!' };
    res.redirect('/blog/admin');
  } catch (error) {
    console.error('Create blog post error:', error);
    req.session.flash = { type: 'danger', message: 'Failed to create blog post' };
    res.redirect('/blog/admin/create');
  }
});

// Edit blog post form
router.get('/admin/edit/:id', requireAdmin, async (req, res) => {
  try {
    const post = await BlogPost.findByPk(req.params.id);
    
    if (!post) {
      req.session.flash = { type: 'danger', message: 'Blog post not found' };
      return res.redirect('/blog/admin');
    }
    
    res.render('admin/blog_post_form', { post });
  } catch (error) {
    console.error('Edit blog post form error:', error);
    req.session.flash = { type: 'danger', message: 'Failed to load blog post' };
    res.redirect('/blog/admin');
  }
});

// Update blog post
router.post('/admin/edit/:id', requireAdmin, async (req, res) => {
  try {
    const post = await BlogPost.findByPk(req.params.id);
    
    if (!post) {
      req.session.flash = { type: 'danger', message: 'Blog post not found' };
      return res.redirect('/blog/admin');
    }
    
    const { title, content, excerpt } = req.body;
    
    if (!title || !content) {
      req.session.flash = { type: 'danger', message: 'Title and content are required' };
      return res.redirect(`/blog/admin/edit/${req.params.id}`);
    }
    
    post.title = title;
    post.content = content;
    post.excerpt = excerpt || content.substring(0, 200) + '...';
    await post.save();
    
    req.session.flash = { type: 'success', message: 'Blog post updated successfully!' };
    res.redirect('/blog/admin');
  } catch (error) {
    console.error('Update blog post error:', error);
    req.session.flash = { type: 'danger', message: 'Failed to update blog post' };
    res.redirect(`/blog/admin/edit/${req.params.id}`);
  }
});

// Delete blog post
router.post('/admin/delete/:id', requireAdmin, async (req, res) => {
  try {
    const post = await BlogPost.findByPk(req.params.id);
    
    if (!post) {
      req.session.flash = { type: 'danger', message: 'Blog post not found' };
      return res.redirect('/blog/admin');
    }
    
    await post.destroy();
    
    req.session.flash = { type: 'success', message: 'Blog post deleted successfully!' };
    res.redirect('/blog/admin');
  } catch (error) {
    console.error('Delete blog post error:', error);
    req.session.flash = { type: 'danger', message: 'Failed to delete blog post' };
    res.redirect('/blog/admin');
  }
});

module.exports = router; 